### PR TITLE
feat: gateway outgoing payments

### PR DESCRIPTION
**Pull request by m1sterc001guy**
State machines and client code for the lightning gateway's outgoing payments

### CODE REVIEW

**Comment by justinmoon** [comment]
looks like there's some issues with rustdoc in ci

**Comment by m1sterc001guy** [comment]
> looks like there's some issues with rustdoc in ci

Yeah had some issues with the aquamarine diagram. Should be fixed now.

File: gateway/ln-gateway/src/ng/mod.rs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+const GW_ANNOUNCEMENT_TTL: Duration = Duration::from_secs(600);
+
+/// The high-level state of a reissue operation started with
+/// [`GatewayClientExt::gateway_pay_bolt11_invoice`].
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum GatewayExtPayStates {
+    Created,
+    Preimage,
+    Success,
```
Comment (justinmoon) [inline gateway/ln-gateway/src/ng/mod.rs:48]: Return `Preimage` here, too. I think this is what the `/pay_invoice` HTTP method will actually be waiting for.

File: modules/fedimint-ln-common/src/api.rs [hunk 2]
```diff
@@ -44,7 +44,7 @@ where
     }
 
     async fn fetch_contract(&self, contract: ContractId) -> FederationResult<ContractAccount> {
-        self.request_current_consensus("wait_account".to_string(), ApiRequestErased::new(contract))
+        self.request_current_consensus("account".to_string(), ApiRequestErased::new(contract))
```
Comment (justinmoon) [inline modules/fedimint-ln-common/src/api.rs:47]: I'm not sure about this change.

Comment (justinmoon) [reply modules/fedimint-ln-common/src/api.rs:47]: Also, the difference between a "contract" and a "contract account" has always been confusing. Not much to do here, but just making note.

Comment (justinmoon) [reply modules/fedimint-ln-common/src/api.rs:47]: This actually seems like a good change. `fetch_contract` should probably just get the contract and fail if it's not there. I'm not aware of any case where the gateway will go looking for a contract before it has been created and need to wait for it to show up.

File: gateway/ln-gateway/src/ng/pay.rs [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+                    common,
+                    state: GatewayPayStates::CancelContract(Box::new(GatewayPayCancelContract {
+                        contract,
+                        error: e,
+                    })),
+                },
+                OutgoingPaymentError::OutgoingContractDoesNotExist { contract_id } => {
+                    GatewayPayStateMachine {
+                        common,
+                        state: GatewayPayStates::Canceled(None, contract_id),
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/ng/pay.rs:259]: `OfferDoesNotExist` sounds better. Will make that change.

File: modules/fedimint-ln-common/src/api.rs (hunk 2 above)
Comment (m1sterc001guy) [reply modules/fedimint-ln-common/src/api.rs:47]: The gateway might not, but does the lightning client? This is shared between both. I don't think this is necessary, I would probably prefer to revert unless others have strong opinions.

File: gateway/ln-gateway/src/ng/tests.rs [hunk 4]
```diff
@@ -110,8 +110,19 @@ async fn test_gateway_client_pay_valid_invoice() -> anyhow::Result<()> {
...
+                _ => {
+                    panic!("Unexpected state occurred, expected Preimage");
+                }
+            }
+            match gw_pay_sub.ok().await? {
+                GatewayExtPayStates::Success { preimage: _ } => {}
+                _ => {
+                    panic!("Unexpected state occurred, expected Success");
+                }
+            }
```
Comment (justinmoon) [inline gateway/ln-gateway/src/ng/tests.rs:124]: I think you can use something like `assert_matches!(gw_pay_sub.ok().await?, GatewayExtPayStates::Success { .. });`

**Pull request merged**